staging/vhost-device-gpu/src/protocol.rs

```diff
@@ -699,7 +702,27 @@ impl GpuCommand {
                 TransferFromHost3d(reader.read_obj().map_err(|_| Error::DescriptorReadFailed)?)
             }
             VIRTIO_GPU_CMD_SUBMIT_3D => {
-                CmdSubmit3d(reader.read_obj().map_err(|_| Error::DescriptorReadFailed)?)
+                let info: virtio_gpu_cmd_submit =
+                    reader.read_obj().map_err(|_| Error::DescriptorReadFailed)?;
+
+                let mut cmd_data = vec![0; info.size as usize];
+                let mut fence_ids: Vec<u64> = Vec::with_capacity(info.num_in_fences as usize);
+
+                for _ in 0..info.num_in_fences {
+                    let fence_id = reader
+                        .read_obj::<u64>()
+                        .map_err(|_| Error::DescriptorReadFailed)?;
+                    fence_ids.push(fence_id);
+                }
+
+                reader
+                    .read_exact(&mut cmd_data[..])
+                    .map_err(|_| Error::DescriptorReadFailed)?;
+
+                CmdSubmit3d {
+                    cmd_data,
+                    fence_ids,
+                }
```

Comment:
I'm not sure what you mean. Do you mean we should handle reading the message in `process_gpu_command` like the comented out code?
I think doing it this way has an advantage as it separates the code nicely for unit testing - we don't need to pass a `Reader` to `process_gpu_command`.  So you can easily test `process_gpu_command` with `CmdSubmit3d` with various data without having to construct a whole descriptor chain, fill it with data and then construct a `Reader`.

Or do you just mean encapsulating the  `cmd_data`, `fence_ids` into a struct and having a function for parsing it?
In that case we might want to do the same with `VIRTIO_GPU_CMD_RESOURCE_ATTACH_BACKING`, which is already merged.

Oh right, I forgot to remove the commented out code.

Reply:
> Do you mean we should handle reading the message in process_gpu_command like the comented out code?

Yes, I meant this. 

> I think doing it this way has an advantage as it separates the code nicely for unit testing - we don't need to pass > a Reader to process_gpu_command. So you can easily test process_gpu_command with CmdSubmit3d with 
> various data without having to construct a whole descriptor chain, fill it with data and then construct a Reader.

How so? in the end when testing this code we will still need to construct a reader because
```               let info: virtio_gpu_cmd_submit =
                    reader.read_obj().map_err(|_| Error::DescriptorReadFailed)?;
```

Reply:
> > I think doing it this way has an advantage as it separates the code nicely for unit testing - we don't need to pass > a Reader to process_gpu_command. So you can easily test process_gpu_command with CmdSubmit3d with
> > various data without having to construct a whole descriptor chain, fill it with data and then construct a Reader.
> 
> How so? in the end when testing this code we will still need to construct a reader because
> 
> ```
>  let info: virtio_gpu_cmd_submit =
>                     reader.read_obj().map_err(|_| Error::DescriptorReadFailed)?;
> ```
Well yes you need to construct a reader, but only for testing the `GpuCommand::decode` for which you already always need a reader, but we curently don't need a `Reader` for `process_gpu_command`. 

To me this seems like a nicer separation of concerns to have  `GpuCommand::decode` does decoding/reading of the whole command instead of just a part of it. And then when executing `process_gpu_command` you don't need a reader anymore - less parameters, easier to unit test

Reply:
> And then when executing process_gpu_command you don't need a reader anymore - less parameters, easier to > unit test
ok makes sense.